### PR TITLE
Output plugin s3

### PR DESCRIPTION
**Pull request by nevanna**

### CODE REVIEW

File: plugin/output/file/file.go [hunk 1]
```diff
@@ -33,6 +33,8 @@ type Plugin struct {
 	fileExtension string
 	fileName      string
 
+	AdditionalFunc func(string) ()
```
Comment (vitkovskii) [inline plugin/output/file/file.go:36]: SealupCallback

File: plugin/output/s3/compress.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	defer zipWriter.Close()
+
+	fileToZip, err := os.Open(fileName)
+	if err != nil {
+		z.logger.Panicf("could not open file: %s, error: %s", fileName, err.Error())
+	}
+	defer fileToZip.Close()
+	info, err := fileToZip.Stat()
+	if err != nil {
+		z.logger.Panicf("could not get file inforamtion for file: %s, error: %s", fileName, err.Error())
```
Comment (vitkovskii) [inline plugin/output/s3/compress.go:48]: `inforamtion` => `information`

File: plugin/output/s3/s3.go [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+	//> File mode for log files
+	FileMode  cfg.Base8 `json:"file_mode" default:"0666" parse:"base8"` //*
+	FileMode_ int64
+
+	//> Compression type
+	CompressionType string `json:"compression_type" default:"zip"` //*
+
+	//	s3 section
+	Endpoint        string `json:"endpoint" required:"true"`
+	AccessKey       string `json:"accessKey" required:"true"`
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:104]: use underscore: `access_key`

File: plugin/output/s3/s3.go [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+	SecretAccessKey string `json:"secretKey" required:"true"`
+	Bucket          string `json:"bucket" required:"true"`
+	Secure          bool   `json:"insecure" default:"false"`
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:107]: Secure or insecure ?))))

File: plugin/output/s3/s3.go [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+	FileMode  cfg.Base8 `json:"file_mode" default:"0666" parse:"base8"` //*
+	FileMode_ int64
+
+	//> Compression type
+	CompressionType string `json:"compression_type" default:"zip"` //*
+
+	//	s3 section
+	Endpoint        string `json:"endpoint" required:"true"`
+	AccessKey       string `json:"accessKey" required:"true"`
+	SecretAccessKey string `json:"secretKey" required:"true"`
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:105]: SecretKey?

File: plugin/output/s3/s3.go [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+	}
+	p.compressor = getNewCompressor(p.logger)
+
+	dir, f := filepath.Split(p.config.TargetFile)
+
+	p.targetDir = dir
+	p.fileExtension = filepath.Ext(f)
+	p.fileName = f[0 : len(f)-len(p.fileExtension)]
+
+	p.uploadCh = make(chan string, p.config.WorkersCount_)
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:140]: it should be different worker count for upload and for compress

File: plugin/output/s3/s3.go [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+	files, err := filepath.Glob(pattern)
+
+	if err != nil {
+		p.logger.Panicf("could not read dir: %s", p.targetDir)
+	}
+	// add sorting
+	//sort files by creation time
+	sort.Slice(files, p.getSortFunc(files))
+	for _, f := range files {
+		p.compressor.compress(p.compressor.getName(f), f)
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:237]: pass compress job to worker

File: plugin/output/s3/s3.go [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+//uploadWork uploads compressed files from channel to s3 and then delete compressed file
+// in case error worker will attempt sending with an exponential time interval
+func (p *Plugin) uploadWork() {
+	for compressed := range p.uploadCh {
+		sleepTime := attemptInterval
+		for {
+			err := p.uploadCompressedFile(compressed)
+			if err == nil {
+				//delete archive after uploading
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:275]: log here successful uploading

File: plugin/output/s3/s3.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+	p.logger = params.Logger
+	p.config = config.(*Config)
+
+	//set up compression
+	p.config.CompressionType = strings.ToLower(p.config.CompressionType)
+	getNewCompressor, ok := compressors[p.config.CompressionType]
+	if !ok {
+		p.logger.Panicf("compression type: %s is not supported", p.config.CompressionType)
+	}
+	p.compressor = getNewCompressor(p.logger)
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:105]: `NewCompressor`

File: plugin/output/s3/s3.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+	p.uploadCh = make(chan string, p.config.FileConfig.WorkersCount_*4)
+	p.compressCh = make(chan string, p.config.FileConfig.WorkersCount_)
+
+	for i := 0; i < p.config.FileConfig.WorkersCount_; i++ {
+		go p.uploadWork()
+		go p.compressWork()
+	}
+
+	//Initialize minio client object.
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:121]: lowercase and space after `//`

File: plugin/output/s3/s3.go [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+func Factory() (pipeline.AnyPlugin, pipeline.AnyConfig) {
+	return &Plugin{}, &Config{}
+}
+
+func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.OutputPluginParams) {
+	p.controller = params.Controller
+	p.logger = params.Logger
+	p.config = config.(*Config)
+
+	//set up compression
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:99]: lowercase and space after `//`

File: plugin/output/s3/s3.go [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+		p.logger.Info("set mock client")
+		p.client = *p.config.client
+	}
+
+	exist, err := p.client.BucketExists(p.config.Bucket)
+	if err != nil {
+		p.logger.Panicf("could not check bucket: %s, error: %s", p.config.Bucket, err.Error())
+	}
+	if !exist {
+		p.logger.Panicf("bucket: %s, does not exist", p.config.Bucket)
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:138]: Error

File: plugin/output/s3/s3.go [hunk 13]
```diff
@@ -0,0 +1,269 @@
...
+	p.logger.Info("client is ready")
+	p.logger.Infof("bucket: %s exists", p.config.Bucket)
+
+	anyPlugin, _ := file.Factory()
+	p.outPlugin = anyPlugin.(*file.Plugin)
+
+	p.outPlugin.SealupCallback = p.addFileJob
+
+	p.outPlugin.Start(&p.config.FileConfig, params)
+	p.uploadExistedFiles()
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:149]: `uploadExistingFiles`

File: plugin/output/s3/s3.go [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+	pattern := fmt.Sprintf("%s*%s", p.targetDir, p.compressor.getExtension())
+	compressedFiles, err := filepath.Glob(pattern)
+	if err != nil {
+		p.logger.Panicf("could not read dir: %s", p.targetDir)
+	}
+	//sort compressed files by creation time
+	sort.Slice(compressedFiles, p.getSortFunc(compressedFiles))
+	//upload archive
+	for _, z := range compressedFiles {
+		p.uploadCh <- z
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:174]: send files further to upload in the compressing stage
if you do it here you'll wait till all files will be compressed and only after that you call uploading stage

File: plugin/output/s3/s3.go [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+
+	if err != nil {
+		p.logger.Panicf("could not read dir: %s", p.targetDir)
+	}
+	//sort files by creation time
+	sort.Slice(files, p.getSortFunc(files))
+	for _, f := range files {
+		p.compressCh <- f
+		//delete old file
+		if err := os.Remove(f); err != nil {
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:191]: it should be in the compression state or you may forget to do it in other place

File: plugin/output/s3/s3.go [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+		}
+		return InfoI.ModTime().Before(InfoJ.ModTime())
+	}
+}
+
+func (p *Plugin) addFileJob(fileName string) {
+	p.compressCh <- fileName
+}
+
+//uploadWork uploads compressed files from channel to s3 and then delete compressed file
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:217]: space

File: plugin/output/s3/s3.go [hunk 17]
```diff
@@ -0,0 +1,269 @@
...
+// in case error worker will attempt sending with an exponential time interval
+func (p *Plugin) uploadWork() {
+	for compressed := range p.uploadCh {
+		sleepTime := attemptInterval
+		for {
+			err := p.uploadCompressedFile(compressed)
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:223]: `uploadToS3`

File: plugin/output/s3/s3.go [hunk 18]
```diff
@@ -0,0 +1,269 @@
...
+			if err == nil {
+				p.logger.Infof("successfully uploaded object: %s", compressed)
+				//delete archive after uploading
+				err = os.Remove(compressed)
+				if err != nil {
+					p.logger.Panicf("could not delete file: %s, err: %s", compressed, err.Error())
+				}
+				break
+			}
+			p.logger.Errorf("could not upload object: %s, error: %s", compressed, err.Error())
```
Comment (vitkovskii) [inline plugin/output/s3/s3.go:233]: log here how much time we'll be wait for next attmpt

**Review by vitkovskii — CHANGES_REQUESTED** [review]

File: plugin/output/s3/s3.go [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+func (p *Plugin) Start(config pipeline.AnyConfig, params *pipeline.OutputPluginParams) {
+	p.controller = params.Controller
+	p.logger = params.Logger
+	p.config = config.(*Config)
+
+	//set up compression
+	p.config.CompressionType = strings.ToLower(p.config.CompressionType)
+	getNewCompressor, ok := compressors[p.config.CompressionType]
+	if !ok {
+		p.logger.Panicf("compression type: %s is not supported", p.config.CompressionType)
```
[the remark this answers is not in the document]
Comment (nevanna) [reply plugin/output/s3/s3.go:103]: Fatalf

**Pull request merged**